File: SAMPLES/JS/classes.js
// classes.js - Exemple JS86 : class, constructor, extends, super, new
// Compile : JS86.EXE classes.js classes.asm

// Classe de base
class Animal {
  constructor(name, sound) {
    this.name = name;
    this.sound = sound;
  }

  speak() {
    console.log(this.sound);
  }
}

// Heritage
class Dog extends Animal {
  constructor(name) {
    super(name, "Woof");
  }
}

class Cat extends Animal {
  constructor(name) {
    super(name, "Meow");
  }
}

// Instanciation
var dog = new Dog("Rex");
dog.speak();
// attendu : Woof

var cat = new Cat("Mimi");
cat.speak();
// attendu : Meow

// Acces aux proprietes
console.log(dog.name);
// attendu : Rex

console.log(cat.name);
// attendu : Mimi

// Classe simple
class Counter {
  constructor() {
    this.count = 0;
  }

  increment() {
    this.count = this.count + 1;
  }

  getCount() {
    return this.count;
  }
}

var c = new Counter();
c.increment();
c.increment();
c.increment();
console.log(c.getCount());
// attendu : 3

File: SAMPLES/PYTHON/classes.py
# classes.py - Exemple PYTHON86 : class, __init__, methodes, heritage
# Compile : PYTHON86.EXE SAMPLES\PYTHON\classes.py classes.asm

# Classe simple
class Counter:
    def __init__(self):
        self.count = 0

    def increment(self):
        self.count = self.count + 1

    def get(self):
        return self.count

print("=== Counter ===")
c = Counter()
print("Initial :", c.get())
c.increment()
c.increment()
c.increment()
print("Apres 3 increments :", c.get())

# Classe avec parametres
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def display(self):
        print("Point(", self.x, ",", self.y, ")")

    def translate(self, dx, dy):
        self.x = self.x + dx
        self.y = self.y + dy

print()
print("=== Point ===")
p = Point(3, 4)
p.display()
p.translate(10, 20)
p.display()

# Heritage
class Animal:
    def __init__(self, nom):
        self.nom = nom

    def parler(self):
        print(self.nom, "fait du bruit")

class Chien(Animal):
    def __init__(self, nom):
        self.nom = nom

    def parler(self):
        print(self.nom, "dit Wouf!")

class Chat(Animal):
    def __init__(self, nom):
        self.nom = nom

    def parler(self):
        print(self.nom, "dit Miaou!")

print()
print("=== Heritage ===")
a = Animal("GenericAnimal")
a.parler()

d = Chien("Rex")
d.parler()

c = Chat("Felix")
c.parler()

# isinstance
print()
print("=== isinstance ===")
if isinstance(d, Chien):
    print("d est un Chien")
if isinstance(d, Animal):
    print("d est un Animal")

print()
print("Fin de classes.py")
